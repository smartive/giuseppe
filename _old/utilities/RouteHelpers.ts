import { VersionInformation } from '../models/VersionInformation';

/**
 * Does compare two route version information to determine if the provided
 * version information is overlapping. Missing parameters are filled with infinity.
 * Algorithm: Not overlapping IF: until-1 < from-2 || from-1 > until-2
 * 
 * @returns {boolean} - true when the two routes are overlapping in some sort, otherwise false.
 */
export function doRouteVersionsOverlap(v1: VersionInformation, v2: VersionInformation): boolean {
    return v1 === v2 || (v1.until >= v2.from && v1.from <= v2.until);
}
