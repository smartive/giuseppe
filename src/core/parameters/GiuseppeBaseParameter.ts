import 'reflect-metadata';

import { Request } from 'express';

import { ParameterParseError, ParameterValidationFailedError, RequiredParameterNotProvidedError } from '../../errors';
import { ParameterDefinition } from '../../parameter/ParameterDefinition';
import { ParameterFactory, ParameterValidator } from './ParameterAdditions';

/**
 * Declaration of primitive javascript types.
 *
 * @type {Function[]}
 */
const PRIMITIVE_TYPES: Function[] = [Object, String, Array, Number, Boolean];

/**
 * Giuseppe base parameter class. Is used for all core parameters (like cookie or body) to generalize
 * some logic.
 *
 * @export
 * @abstract
 * @class GiuseppeBaseParameter
 * @implements {ParameterDefinition}
 */
export abstract class GiuseppeBaseParameter implements ParameterDefinition {
    public readonly canHandleResponse: boolean = false;

    constructor(
        public readonly name: string,
        public readonly type: Function,
        public readonly index: number,
        protected readonly required?: boolean,
        protected readonly validator?: ParameterValidator,
        protected readonly factory?: ParameterFactory<any>,
    ) { }

    public getValue(request: Request): any {
        let value = this.getRawValue(request);
        value = this.parseValue(value);
        this.validateValue(value);
        return value;
    }

    protected abstract getRawValue(request: Request): any;

    protected parseValue(raw: any): any {
        if (!raw) {
            return raw;
        }

        let factory;
        if (this.factory) {
            factory = this.factory;
        } else {
            factory = rawValue => {
                const ctor = this.type as any;
                if (rawValue.constructor === ctor) {
                    return rawValue;
                }
                return PRIMITIVE_TYPES.indexOf(ctor) !== -1 ? ctor(rawValue) : new ctor(rawValue);
            };
        }

        try {
            return factory(raw);
        } catch (e) {
            throw new ParameterParseError(this.name, e);
        }
    }

    protected validateValue(parsed: any): void {
        if ((parsed === null || parsed === undefined) && this.required) {
            throw new RequiredParameterNotProvidedError(this.name);
        }

        if (!this.validator || (!this.required && (parsed === undefined || parsed === null))) {
            return;
        }
        const isValid = value => {
            const predicates = this.validator;

            if (Array.isArray(predicates)) {
                return predicates.every(p => p(value));
            }

            return predicates!(value);
        };

        try {
            if (!isValid(parsed)) {
                throw new ParameterValidationFailedError(this.name);
            }
        } catch (e) {
            if (e instanceof ParameterValidationFailedError) {
                throw e;
            }
            let error: Error;
            if (!(e instanceof Error)) {
                error = new Error(e);
            } else {
                error = e;
            }
            throw new ParameterValidationFailedError(this.name, error);
        }
    }
}
