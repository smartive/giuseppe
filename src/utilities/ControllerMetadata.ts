import 'reflect-metadata';

import { ControllerErrorHandler } from '../errors/ControllerErrorHandler';
import { ERRORHANDLER_KEY } from '../errors/ErrorHandlerDecorator';
import { PARAMETER_DEFINITION_KEY, ParameterDefinition } from '../parameter/ParameterDefinition';
import { ROUTE_DEFINITION_KEY, RouteDefinition } from '../routes/RouteDefinition';
import { ROUTE_MODIFICATOR_KEY, RouteModificator } from '../routes/RouteModificator';

/**
 * Helper class that manages metadata elements on a controller class.
 * 
 * @export
 * @class ControllerMetadata
 */
export class ControllerMetadata {

    constructor(private controller: Function | Object) { }

    /**
     * Returns the defined routes of a controller class. If no routes are defined, defines the array of routes
     * on the controller and returns the empty array.
     * 
     * @returns {RouteDefinition[]} 
     * @memberof ControllerMetadata
     */
    public routes(): RouteDefinition[] {
        if (!Reflect.hasOwnMetadata(ROUTE_DEFINITION_KEY, this.controller)) {
            Reflect.defineMetadata(ROUTE_DEFINITION_KEY, [], this.controller);
        }
        return Reflect.getOwnMetadata(ROUTE_DEFINITION_KEY, this.controller);
    }

    /**
     * Returns the defined route modificators for a given route name of a controller class.
     * If no modificators are defined, defines the array of modificators on the controller and returns the empty array.
     * 
     * @param {string} name 
     * @returns {RouteModificator[]} 
     * @memberof ControllerMetadata
     */
    public modificators(name: string): RouteModificator[] {
        if (!Reflect.hasOwnMetadata(ROUTE_MODIFICATOR_KEY, this.controller, name)) {
            Reflect.defineMetadata(ROUTE_MODIFICATOR_KEY, [], this.controller, name);
        }
        return Reflect.getOwnMetadata(ROUTE_MODIFICATOR_KEY, this.controller, name);
    }

    /**
     * Returns the defined route parameters for a given route name of a controller class.
     * If no parameters are defined, defines the array of parameters on the controller and returns the empty array.
     * 
     * @param {string} name 
     * @returns {ParameterDefinition[]} 
     * @memberof ControllerMetadata
     */
    public parameters(name: string): ParameterDefinition[] {
        if (!Reflect.hasOwnMetadata(PARAMETER_DEFINITION_KEY, this.controller, name)) {
            Reflect.defineMetadata(PARAMETER_DEFINITION_KEY, [], this.controller, name);
        }
        return Reflect.getOwnMetadata(PARAMETER_DEFINITION_KEY, this.controller, name);
    }

    /**
     * Returns the types for the parameters of a given route name. Returns an empty array if the route has no parameters.
     * 
     * @param {string} name 
     * @returns {Function[]} 
     * @memberof ControllerMetadata
     */
    public parameterTypes(name: string): Function[] {
        return Reflect.getOwnMetadata('design:paramtypes', this.controller, name) || [];
    }

    /**
     * Returns the defined error handler of a controller class. If no handler is defined, defines the empty handler
     * on the controller and returns it.
     * 
     * @returns {ControllerErrorHandler} 
     * @memberof ControllerMetadata
     */
    public errorHandler(): ControllerErrorHandler {
        if (!Reflect.hasOwnMetadata(ERRORHANDLER_KEY, this.controller)) {
            Reflect.defineMetadata(ERRORHANDLER_KEY, new ControllerErrorHandler(), this.controller);
        }
        return Reflect.getOwnMetadata(ERRORHANDLER_KEY, this.controller);
    }
}
