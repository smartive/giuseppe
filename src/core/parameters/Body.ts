import 'reflect-metadata';
import { Giuseppe } from '../../Giuseppe';
import { ControllerMetadata } from '../../utilities/ControllerMetadata';
import { GiuseppeBaseParameter } from './GiuseppeBaseParameter';
import { ParameterFactory, ParameterValidator } from './ParameterAdditions';
import { Request } from 'express';

export function Body(
    {
        validator,
        factory,
    }: { validator?: ParameterValidator, factory?: ParameterFactory<any> } = {},
): ParameterDecorator {
    return (target: Object, propertyKey: string, parameterIndex: number) =>
        Giuseppe.registrar.registerParameter(
            target,
            propertyKey,
            new GiuseppeBodyParameter(
                new ControllerMetadata(target).parameterTypes(propertyKey)[parameterIndex],
                parameterIndex,
                true,
                validator,
                factory,
            ),
        );
}

export class GiuseppeBodyParameter extends GiuseppeBaseParameter {

    constructor(
        type: Function,
        index: number,
        required?: boolean,
        validator?: ParameterValidator,
        factory?: ParameterFactory<any>,
    ) {
        super('body', type, index, required, validator, factory);
    }

    protected getRawValue(request: Request): any {
        return request.body;
    }
}
