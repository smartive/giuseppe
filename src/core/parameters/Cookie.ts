import 'reflect-metadata';

import { Request } from 'express';

import { Giuseppe } from '../../Giuseppe';
import { ControllerMetadata } from '../../utilities/ControllerMetadata';
import { GiuseppeBaseParameter } from './GiuseppeBaseParameter';
import { ParameterFactory, ParameterValidator } from './ParameterAdditions';

/**
 * Class that splits cookie values into their specific parts.
 *
 * @class CookieHelper
 */
class CookieHelper {
  public name: string;
  public value: string;

  constructor(value: string) {
    const split = value.split('=');
    this.name = split[0];
    this.value = split[1];
  }
}

/**
 * Parameter decorator. Creates a parameter definition that injects a specific cookie from the request.
 * Can contain validators and a factory if the value is complex.
 *
 * @export
 * @param {string} name Name of the cookie.
 * @param {{ required?: boolean, validator?: ParameterValidator, factory?: ParameterFactory<any> }} [{
 *         required,
 *         validator,
 *         factory,
 *     }={}] Configuration object for the cookie parameter.
 * @returns {ParameterDecorator}
 */
export function Cookie(
  name: string,
  {
    required,
    validator,
    factory,
  }: { required?: boolean; validator?: ParameterValidator; factory?: ParameterFactory<any> } = {},
): ParameterDecorator {
  return (target: Object, propertyKey: string | symbol, parameterIndex: number) =>
    Giuseppe.registrar.registerParameter(
      target,
      propertyKey.toString(),
      new GiuseppeCookieParameter(
        name,
        new ControllerMetadata(target).parameterTypes(propertyKey.toString())[parameterIndex],
        parameterIndex,
        required,
        validator,
        factory,
      ),
    );
}

/**
 * Default core cookie parameter of giuseppe. Injects the value of a specific cookie from the request.
 *
 * @export
 * @class GiuseppeCookieParameter
 * @extends {GiuseppeBaseParameter}
 */
export class GiuseppeCookieParameter extends GiuseppeBaseParameter {
  protected getRawValue(request: Request): any {
    const cookies = request.get('cookie');
    if (!cookies) {
      return undefined;
    }
    const cookie = cookies
      .split(';')
      .map((o) => new CookieHelper(o.trim()))
      .filter((o) => o.name === this.name)[0];
    return cookie ? cookie.value : undefined;
  }
}
