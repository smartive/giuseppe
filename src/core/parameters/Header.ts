import 'reflect-metadata';
import { Giuseppe } from '../../Giuseppe';
import { ControllerMetadata } from '../../utilities/ControllerMetadata';
import { GiuseppeBaseParameter } from './GiuseppeBaseParameter';
import { ParameterFactory, ParameterValidator } from './ParameterAdditions';
import { Request } from 'express';

export function Header(
    name: string,
    {
        required,
        validator,
        factory,
    }: { required?: boolean, validator?: ParameterValidator, factory?: ParameterFactory<any> } = {},
): ParameterDecorator {
    return (target: Object, propertyKey: string, parameterIndex: number) =>
        Giuseppe.registrar.registerParameter(
            target,
            propertyKey,
            new GiuseppeHeaderParameter(
                name,
                new ControllerMetadata(target).parameterTypes(propertyKey)[parameterIndex],
                parameterIndex,
                required,
                validator,
                factory,
            ),
        );
}

export class GiuseppeHeaderParameter extends GiuseppeBaseParameter {

    constructor(
        name: string,
        type: Function,
        index: number,
        required?: boolean,
        validator?: ParameterValidator,
        factory?: ParameterFactory<any>,
    ) {
        super(name, type, index, required, validator, factory);
    }

    protected getRawValue(request: Request): any {
        return request.get(this.name);
    }
}
