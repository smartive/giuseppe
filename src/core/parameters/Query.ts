import 'reflect-metadata';

import { Request } from 'express';

import { Giuseppe } from '../../Giuseppe';
import { ControllerMetadata } from '../../utilities/ControllerMetadata';
import { GiuseppeBaseParameter } from './GiuseppeBaseParameter';
import { ParameterFactory, ParameterValidator } from './ParameterAdditions';

/**
 * @typedef QueryConfigObject

 * @property {string | string[]} [alias]
 * @property {boolean} [required]
 * @property {ParameterValidator} [validator]
 * @property {ParameterFactory<any>} [factory]
 */

/**
 * Parameter decorator. Creates a parameter definition that injects a query parameter value from the request.
 * Can contain validators and a factory if the value is complex. If an alias is defined, the alias(es)
 * will be used to determine a value other than the given name.
 *
 * @export
 * @param {string} name
 * @param {QueryConfigObject} [{alias, required, validator, factory}={}]
 * @returns {ParameterDecorator}
 */
export function Query(
  name: string,
  {
    alias,
    required,
    validator,
    factory,
  }: {
    alias?: string | string[];
    required?: boolean;
    validator?: ParameterValidator;
    factory?: ParameterFactory<any>;
  } = {},
): ParameterDecorator {
  return (target: Object, propertyKey: string | symbol, parameterIndex: number) =>
    Giuseppe.registrar.registerParameter(
      target,
      propertyKey.toString(),
      new GiuseppeQueryParameter(
        name,
        new ControllerMetadata(target).parameterTypes(propertyKey.toString())[parameterIndex],
        parameterIndex,
        required,
        validator,
        factory,
        alias,
      ),
    );
}

/**
 * Default core query parameter of giuseppe. Does inject a query parameter from the request into the route.
 * Can contain an alias that the query parameter can be named with.
 *
 * @export
 * @class GiuseppeQueryParameter
 * @extends {GiuseppeBaseParameter}
 */
export class GiuseppeQueryParameter extends GiuseppeBaseParameter {
  constructor(
    name: string,
    type: Function,
    index: number,
    required?: boolean,
    validator?: ParameterValidator,
    factory?: ParameterFactory<any>,
    private readonly alias?: string | string[],
  ) {
    super(name, type, index, required, validator, factory);
  }

  protected getRawValue(request: Request): any {
    if (!this.alias) {
      return request.query[this.name];
    }
    let aliases = !Array.isArray(this.alias) ? [this.alias] : (this.alias as string[]);
    aliases = aliases.map((a: string) => request.query[a]).filter(Boolean);
    return aliases[0] || request.query[this.name];
  }
}
