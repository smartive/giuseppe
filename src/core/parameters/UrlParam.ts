import { Request } from 'express';
import 'reflect-metadata';
import { Giuseppe } from '../../Giuseppe';
import { ControllerMetadata } from '../../utilities/ControllerMetadata';
import { GiuseppeBaseParameter } from './GiuseppeBaseParameter';
import { ParameterFactory, ParameterValidator } from './ParameterAdditions';

/**
 * Parameter decorator. Creates a parameter definition that injects a url parameter value from the request.
 * Can contain validators and a factory if the value is complex.
 *
 * @export
 * @param {string} name
 * @param {{ validator?: ParameterValidator, factory?: ParameterFactory<any> }} [{
 *         validator,
 *         factory,
 *     }={}]
 * @returns {ParameterDecorator}
 */
export function UrlParam(
  name: string,
  { validator, factory }: { validator?: ParameterValidator; factory?: ParameterFactory<any> } = {},
): ParameterDecorator {
  return (target: Object, propertyKey: string | symbol, parameterIndex: number) =>
    Giuseppe.registrar.registerParameter(
      target,
      propertyKey.toString(),
      new GiuseppeUrlParameter(
        name,
        new ControllerMetadata(target).parameterTypes(propertyKey.toString())[parameterIndex],
        parameterIndex,
        validator,
        factory,
      ),
    );
}

/**
 * Default core url parameter of giuseppe. Injects a given url parameter for a route.
 *
 * @export
 * @class GiuseppeUrlParameter
 * @extends {GiuseppeBaseParameter}
 */
export class GiuseppeUrlParameter extends GiuseppeBaseParameter {
  constructor(name: string, type: Function, index: number, validator?: ParameterValidator, factory?: ParameterFactory<any>) {
    super(name, type, index, true, validator, factory);
  }

  protected getRawValue(request: Request): any {
    return request.params[this.name];
  }
}
