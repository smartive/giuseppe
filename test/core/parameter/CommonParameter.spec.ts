import { Cookie, GiuseppeCookieParameter } from '../../../src/core/parameters/Cookie';
import { GiuseppeHeaderParameter, Header } from '../../../src/core/parameters/Header';
import 'reflect-metadata';
import { Body, GiuseppeBodyParameter } from '../../../src/core/parameters/Body';
import { GiuseppeBaseParameter } from '../../../src/core/parameters/GiuseppeBaseParameter';
import {
    isNumber,
    isString,
    ParameterFactory,
    ParameterValidator
} from '../../../src/core/parameters/ParameterAdditions';
import { GiuseppeQueryParameter, Query } from '../../../src/core/parameters/Query';
import { GiuseppeUrlParameter, UrlParam } from '../../../src/core/parameters/UrlParam';
import {
    ParameterParseError,
    ParameterValidationFailedError,
    RequiredParameterNotProvidedError
} from '../../../src/errors';
import { ControllerMetadata } from '../../../src/utilities/ControllerMetadata';

interface ParamCase {
    class: new (...args: any[]) => GiuseppeBaseParameter;
    name: string;
    alwaysRequired: boolean;
    getRequestMock: (value?: any) => any;
    getDecorator: (name?: any) => any;
    getInstance: (
        name: string,
        type: Function,
        index: number,
        required?: boolean,
        validator?: ParameterValidator,
        factory?: ParameterFactory<any>,
    ) => GiuseppeBaseParameter;
}

describe('Giuseppe parameter common', () => {

    const params: ParamCase[] = [
        {
            class: GiuseppeQueryParameter,
            name: 'Query',
            alwaysRequired: false,
            getRequestMock: value => ({ query: { 'name': value } }),
            getDecorator: name => Query(name),
            getInstance: (
                name: string,
                type: Function,
                index: number,
                required?: boolean,
                validator?: ParameterValidator,
                factory?: ParameterFactory<any>,
            ) => new GiuseppeQueryParameter(name, type, index, required, validator, factory),
        },
        {
            class: GiuseppeUrlParameter,
            name: 'UrlParam',
            alwaysRequired: true,
            getRequestMock: value => ({ params: { 'name': value } }),
            getDecorator: name => UrlParam(name),
            getInstance: (
                name: string,
                type: Function,
                index: number,
                required?: boolean,
                validator?: ParameterValidator,
                factory?: ParameterFactory<any>,
            ) => new GiuseppeUrlParameter(name, type, index, validator, factory),
        },
        {
            class: GiuseppeBodyParameter,
            name: 'Body',
            alwaysRequired: false,
            getRequestMock: value => ({ body: value }),
            getDecorator: name => Body(),
            getInstance: (
                name: string,
                type: Function,
                index: number,
                required?: boolean,
                validator?: ParameterValidator,
                factory?: ParameterFactory<any>,
            ) => new GiuseppeBodyParameter(type, index, required, validator, factory),
        },
        {
            class: GiuseppeHeaderParameter,
            name: 'Header',
            alwaysRequired: false,
            getRequestMock: value => ({
                get: name => {
                    const foo = { 'name': value };
                    return foo[name];
                },
            }),
            getDecorator: name => Header(name),
            getInstance: (
                name: string,
                type: Function,
                index: number,
                required?: boolean,
                validator?: ParameterValidator,
                factory?: ParameterFactory<any>,
            ) => new GiuseppeHeaderParameter(name, type, index, required, validator, factory),
        },
        {
            class: GiuseppeCookieParameter,
            name: 'Cookie',
            alwaysRequired: false,
            getRequestMock: value => ({
                get: () => value ? `name=${value};` : null,
            }),
            getDecorator: name => Cookie(name),
            getInstance: (
                name: string,
                type: Function,
                index: number,
                required?: boolean,
                validator?: ParameterValidator,
                factory?: ParameterFactory<any>,
            ) => new GiuseppeCookieParameter(name, type, index, required, validator, factory),
        },
    ];

    for (const param of params) {

        describe(`@${param.name} decorator`, () => {

            it('should return a param decorator', () => {
                expect(param.getDecorator('name')).toBeInstanceOf(Function);
                expect(param.getDecorator('name').length).toBe(3);
            });

            it('should add the correct parameter declaration to the class', () => {
                class Ctrl {
                    public func( @param.getDecorator('name') param: string): void { }
                }

                const meta = new ControllerMetadata(Ctrl.prototype);

                const ctrlParam = meta.parameters('func')[0];

                expect(ctrlParam).toBeInstanceOf(param.class);
            });

            it('should add the correct parameter type to the declaration', () => {
                class Foobar { }

                class Ctrl {
                    public func( @param.getDecorator('name') param: Foobar): void { }
                }

                const meta = new ControllerMetadata(Ctrl.prototype);

                const ctrlParam = meta.parameters('func')[0];

                expect(ctrlParam.type).toBe(Foobar);
            });

        });

        describe(`Giuseppe${param.name}Parameter class`, () => {

            it('should inject the correct value', () => {
                const instance = param.getInstance('name', String, 0);

                expect(instance.getValue(param.getRequestMock('value'))).toBe('value');
            });

            it('should parse the correct value type', () => {
                const instance = param.getInstance('name', String, 0);
                expect(instance.getValue(param.getRequestMock('value')).constructor).toBe(String);

                const instance2 = param.getInstance('name', Number, 0);
                expect(instance2.getValue(param.getRequestMock(1337)).constructor).toBe(Number);
            });

            it('should throw on not provided required parameter', () => {
                const instance = param.getInstance('name', String, 0, true),
                    fn = () => instance.getValue(param.getRequestMock());

                expect(fn).toThrow(RequiredParameterNotProvidedError);
            });

            if (!param.alwaysRequired) {
                it('should inject undefined if param is not provided', () => {
                    const instance = param.getInstance('name', String, 0);

                    expect(instance.getValue(param.getRequestMock())).toBeUndefined();
                });
            }

            it('should correctly parse a custom type', () => {
                class Foobar {
                    constructor(value: any) { }
                }

                const instance = param.getInstance('name', Foobar, 0);

                expect(instance.getValue(param.getRequestMock('value'))).toBeInstanceOf(Foobar);
            });

            it('should throw on a param parse error', () => {
                class Foobar {
                    constructor(value: any) {
                        throw new Error();
                    }
                }

                const instance = param.getInstance('name', Foobar, 0),
                    fn = () => instance.getValue(param.getRequestMock('value'));

                expect(fn).toThrow(ParameterParseError);
            });

            it('should parse correctly with a factory', () => {
                class Foobar {
                    public test: string;

                    private constructor() { }

                    public static create(value: any): Foobar {
                        const f = new Foobar();
                        f.test = value;
                        return f;
                    }
                }

                const instance = param.getInstance('name', String, 0, false, null, raw => Foobar.create(raw));

                expect(instance.getValue(param.getRequestMock('value'))).toBeInstanceOf(Foobar);
                expect(instance.getValue(param.getRequestMock('value')).test).toBe('value');
            });

            it('should throw correctly with a factory', () => {
                class Foobar {
                    public test: string;

                    private constructor() { }

                    public static create(value: any): Foobar {
                        throw new Error();
                    }
                }

                const instance = param.getInstance('name', String, 0, false, null, raw => Foobar.create(raw)),
                    fn = () => instance.getValue(param.getRequestMock('value'));

                expect(fn).toThrow();
            });

            it('should validate correctly', () => {
                const instance = param.getInstance('name', String, 0, false, isString());

                const fn = () => instance.getValue(param.getRequestMock('value'));

                expect(fn).not.toThrow(ParameterValidationFailedError);
            });

            it('should validate correctly with multiple validators', () => {
                const isNotEmpty = value => value.length > 0;

                const instance = param.getInstance('name', String, 0, false, [isString(), isNotEmpty]);

                const fn = () => instance.getValue(param.getRequestMock('value'));

                expect(fn).not.toThrow(ParameterValidationFailedError);
            });

            it('should throw on validation error', () => {
                const instance = param.getInstance('name', Number, 0, false, isString());

                const fn = () => instance.getValue(param.getRequestMock(1337));

                expect(fn).toThrow(ParameterValidationFailedError);
            });

            it('should throw on a later validation error (if multiple)', () => {
                const instance = param.getInstance('name', String, 0, false, [isString(), isNumber()]);

                const fn = () => instance.getValue(param.getRequestMock('1337'));

                expect(fn).toThrow(ParameterValidationFailedError);
            });

        });

    }

});
