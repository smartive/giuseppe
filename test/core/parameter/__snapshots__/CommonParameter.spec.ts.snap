// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Giuseppe parameter common GiuseppeBodyParameter class should throw a validator failed error (with inner exception) if the validator throws a non error 1`] = `
"The validator for the parameter \\"body\\" was not valid.
The validator threw an error: Error: non normal error"
`;

exports[`Giuseppe parameter common GiuseppeBodyParameter class should throw a validator failed error (with inner exception) when the validator throws an error 1`] = `
"The validator for the parameter \\"body\\" was not valid.
The validator threw an error: Error: normal error"
`;

exports[`Giuseppe parameter common GiuseppeBodyParameter class should throw a validator failed error (without inner exception) when the validator throws a validator failed error 1`] = `"The validator for the parameter \\"Body\\" was not valid."`;

exports[`Giuseppe parameter common GiuseppeCookieParameter class should throw a validator failed error (with inner exception) if the validator throws a non error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: non normal error"
`;

exports[`Giuseppe parameter common GiuseppeCookieParameter class should throw a validator failed error (with inner exception) when the validator throws an error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: normal error"
`;

exports[`Giuseppe parameter common GiuseppeCookieParameter class should throw a validator failed error (without inner exception) when the validator throws a validator failed error 1`] = `"The validator for the parameter \\"Cookie\\" was not valid."`;

exports[`Giuseppe parameter common GiuseppeHeaderParameter class should throw a validator failed error (with inner exception) if the validator throws a non error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: non normal error"
`;

exports[`Giuseppe parameter common GiuseppeHeaderParameter class should throw a validator failed error (with inner exception) when the validator throws an error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: normal error"
`;

exports[`Giuseppe parameter common GiuseppeHeaderParameter class should throw a validator failed error (without inner exception) when the validator throws a validator failed error 1`] = `"The validator for the parameter \\"Header\\" was not valid."`;

exports[`Giuseppe parameter common GiuseppeQueryParameter class should throw a validator failed error (with inner exception) if the validator throws a non error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: non normal error"
`;

exports[`Giuseppe parameter common GiuseppeQueryParameter class should throw a validator failed error (with inner exception) when the validator throws an error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: normal error"
`;

exports[`Giuseppe parameter common GiuseppeQueryParameter class should throw a validator failed error (without inner exception) when the validator throws a validator failed error 1`] = `"The validator for the parameter \\"Query\\" was not valid."`;

exports[`Giuseppe parameter common GiuseppeUrlParamParameter class should throw a validator failed error (with inner exception) if the validator throws a non error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: non normal error"
`;

exports[`Giuseppe parameter common GiuseppeUrlParamParameter class should throw a validator failed error (with inner exception) when the validator throws an error 1`] = `
"The validator for the parameter \\"name\\" was not valid.
The validator threw an error: Error: normal error"
`;

exports[`Giuseppe parameter common GiuseppeUrlParamParameter class should throw a validator failed error (without inner exception) when the validator throws a validator failed error 1`] = `"The validator for the parameter \\"UrlParam\\" was not valid."`;
